group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -875,30 +875,43 @@ private void maybeRemovePartitionEpoch(
         ConsumerGroupMember oldMember
     ) {
         if (oldMember != null) {
-            removePartitionEpochs(oldMember.assignedPartitions());
-            removePartitionEpochs(oldMember.partitionsPendingRevocation());
+            removePartitionEpochs(oldMember.assignedPartitions(), oldMember.memberEpoch());
+            removePartitionEpochs(oldMember.partitionsPendingRevocation(), oldMember.memberEpoch());
         }
     }
 
     /**
      * Removes the partition epochs based on the provided assignment.
      *
      * @param assignment    The assignment.
+     * @param expectedEpoch The expected epoch.
+     * @throws IllegalStateException if the epoch does not match the expected one.
+     * package-private for testing.
      */
-    private void removePartitionEpochs(
-        Map<Uuid, Set<Integer>> assignment
+    void removePartitionEpochs(
+        Map<Uuid, Set<Integer>> assignment,
+        int expectedEpoch
     ) {
         assignment.forEach((topicId, assignedPartitions) -> {
             currentPartitionEpoch.compute(topicId, (__, partitionsOrNull) -> {
                 if (partitionsOrNull != null) {
-                    assignedPartitions.forEach(partitionsOrNull::remove);
+                    assignedPartitions.forEach(partitionId -> {
+                        Integer prevValue = partitionsOrNull.remove(partitionId);
+                        if (prevValue != expectedEpoch) {
+                            throw new IllegalStateException(
+                                String.format("Cannot remove the epoch %d from %s-%s because the partition is " +
+                                    "still owned at a different epoch %d", expectedEpoch, topicId, partitionId, prevValue));
+                        }
+                    });
                     if (partitionsOrNull.isEmpty()) {
                         return null;
                     } else {
                         return partitionsOrNull;
                     }
                 } else {
-                    return null;
+                    throw new IllegalStateException(
```

Comment:
What is the affect of throwing this error? Do we also block removing the rest of the partitions?

Reply:
It will fail the entire write operation so we would rollback the entire state to the state of the last written offset. Let's take the ConsumerGroupHeartbeat as an example.

Here is the common flow:
1) A write operation is create to handle the request.
2) The operation handles the request, this generate the response and the records to apply to the state machine.
3) The records are replayed to the state machine.
4) The records are written to the log.
5) The response is returned when the records are replicated.

The exception here would basically break step 3) so the entire operation would be failed and the HB would return an unknown server error (IllegalStateException). When the operation fails, the state machine is rolled back to the state before the operation.

Reply:
Thanks for stepping through this. So the operation/request would need to be retried? Do we expect it to recover on its own when metadata updates for example?

Reply:
The operation would fail and the client would decide what to do. In this case, we would not recover because we would likely hit the same bug again. I suppose that we could recover in different situations. Overall, the point of the change is to fail when the state is incorrect instead of silently accepting it. This will help us to detect those issues.

Reply:
Ack -- so we would essentially get stuck and this would 
1) prevent further issues
2) probably alert the user as the consumer could not continue